internal/restore/volumesnapshot_restorer.go

```diff
@@ -58,16 +62,61 @@ func (p *VSRestorer) Execute(input *velero.RestoreItemActionExecuteInput) (*vele
 		return &velero.RestoreItemActionExecuteOutput{}, errors.Wrapf(err, "failed to convert input.Item from unstructured")
 	}
 
-	var vsFromBackup snapshotv1beta1api.VolumeSnapshot
-	if err := runtime.DefaultUnstructuredConverter.FromUnstructured((input.ItemFromBackup.UnstructuredContent()), &vsFromBackup); err != nil {
-		return &velero.RestoreItemActionExecuteOutput{}, errors.Wrapf(err, "failed to convert input.ItemFromBackup from unstructured")
+	labels := vs.Labels
+	if labels == nil {
+		labels = make(map[string]string)
+	}
+	snapHandle, exists := labels[util.VolumeSnapshotHandleLabel]
+	if !exists {
+		return nil, errors.Errorf("Volumesnapshot %s/%s does not have a %s label", vs.Namespace, vs.Name, util.VolumeSnapshotHandleLabel)
+	}
+	// TODO: refactor to reduce copy-pasta
+	csiDriverName, exists := labels[util.CSIDriverNameLabel]
+	if !exists {
+		return nil, errors.Errorf("Volumesnapshot %s/%s does not have a %s label", vs.Namespace, vs.Name, util.CSIDriverNameLabel)
+	}
+
+	// TODO: generated name will be like velero-velero-something. Fix that.
+	vsc := snapshotv1beta1api.VolumeSnapshotContent{
+		ObjectMeta: metav1.ObjectMeta{
+			GenerateName: "velero-" + vs.Name + "-",
+			Annotations: map[string]string{
+				velerov1api.RestoreNameLabel: input.Restore.Name,
+			},
+		},
+		Spec: snapshotv1beta1api.VolumeSnapshotContentSpec{
+			DeletionPolicy: snapshotv1beta1api.VolumeSnapshotContentDelete,
+			Driver:         csiDriverName,
+			VolumeSnapshotRef: core_v1.ObjectReference{
+				Kind:      "VolumeSnapshot",
+				Namespace: vs.Namespace,
+				Name:      vs.Name,
+			},
+			Source: snapshotv1beta1api.VolumeSnapshotContentSource{
+				SnapshotHandle: &snapHandle,
+			},
+		},
+	}
+
+	// Set Delete snapshot secret annotations if it was present during backup of the dynamic volumesnapshot
+	if util.IsVolumeSnapshotHasVSCDeleteSecret(&vs) {
+		vsc.Annotations[util.PrefixedSnapshotterSecretNameKey] = vs.Annotations[util.CSIDeleteSnapshotSecretName]
+		vsc.Annotations[util.PrefixedSnapshotterSecretNamespaceKey] = vs.Annotations[util.CSIDeleteSnapshotSecretNamespace]
 	}
 
-	if vsFromBackup.Status == nil || vsFromBackup.Status.BoundVolumeSnapshotContentName == nil {
-		return &velero.RestoreItemActionExecuteOutput{}, errors.Errorf("unable to lookup BoundVolumeSnapshotContentName from status")
+	_, snapClient, err := util.GetClients()
+	if err != nil {
+		return nil, errors.WithStack(err)
+	}
+	vscupd, err := snapClient.SnapshotV1beta1().VolumeSnapshotContents().Create(&vsc)
```

Comment:
Did you consider returning the VSC as an additional item to restore, rather than directly creating it within the plugin? That's how we handle PVC->PV restores (see in-tree Velero plugins in pkg/restore). That's probably simpler, though I may be missing something.

Reply:
The logic here is mainly converting the dynamic volumesnapshot to a static volumesnapshot.
Creating a new volumesnapshotcontent is what allows both originally/dynamically bound vsc and statically bound vsc to co-exist in the cluster.
Creating the vsc here also allows to setup the link between the volumesnapshot and volumesnapshotcontent. L118

Reply:
I would have expected the creation of the new, static VSC to happen on restore. Velero's philosophy has usually been to back up things in the cluster more or less as-is, with minimal changes, then do any sort of manipulation/recreation on the restore side. That way, we have exactly what was in the cluster at backup time (even if it's something we created).